ext/src/ruby_api/store.rs

```diff
@@ -95,6 +95,51 @@ impl Store {
     }
 }
 
+/// A wrapper around a Ruby Value that has a store context.
+/// Used in places where both Store or Caller can be used.
+#[derive(Debug)]
+pub enum StoreContextValue {
+    Store(Value),
+    Caller(Value),
+}
+impl StoreContextValue {
+    pub fn mark(&self) {
+        match self {
+            Self::Store(v) => magnus::gc::mark(v),
+            Self::Caller(_) => {
+                // The Caller is on the stack while it's "live". Right before the end of a host call,
+                // we remove the Caller form the Ruby object, thus there is no need to mark.
+            }
+        }
+    }
+
+    pub fn context(&self) -> Result<StoreContext<StoreData>, Error> {
+        match self {
+            Self::Store(s) => {
+                let store = s.try_convert::<&Store>().expect("a Store typed data");
```

Comment:
Would it be better to:

```suggestion
                let store = s.try_convert::<&Store>()?;
```

Reply:
I initially went with `expect` to express that this is an invariant, regardless of what the Ruby program does.

The counter-argument is that we should avoid panic if we can 🤷. No strong opinion on my end, happy to defer to your Rust experience here.